---
title: "scoary_analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r packages, include=TRUE}
library(devtools) 
library(ggplot2)
library(tools)
library(purrr)
library(dplyr)
library(fuzzyjoin)
library(tidyr)
```


```{r data,include=TRUE}
setwd("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/Scoary_Results/")
dataFiles <- sapply(Sys.glob("./*.csv"), read.csv, simplify = FALSE, USE.NAMES = TRUE)

gene_pres_abs <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/gene_presence_absence_roary.csv")

blastp <- read.table("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/ecoli_all_unique.out",header=FALSE)
colnames(blastp) <- c("qseqid","sseqid","sacc","sgi","qaccver","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore")
gene_key <- blastp[,c(1,3)]  
AMR_genes_NCBI <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/AMRFinder_known_genes_NCBI_3.1.21.tsv",header=TRUE,sep="\t")
write.csv(gene_key, file="/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/gene_key_all.csv")
# gene_key <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/gi_gene_key_all_acc.csv",header=TRUE)

# read in giant matrix of gene data 
# gene_data <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/gene_data.csv")


# get the refseq protein and refseq nucleotide IDs into a new dataframe by themselves, without the .1
AMR_genes_NCBI_refseq <- cbind(AMR_genes_NCBI$refseq.protein,AMR_genes_NCBI$refseq.nucleotide)
class(AMR_genes_NCBI_refseq)
AMR_genes_NCBI_refseq <- as.data.frame(AMR_genes_NCBI_refseq)
colnames(AMR_genes_NCBI_refseq) <- c("refSeqProt","refSeqNuc")
AMR_genes_NCBI_refSeqProt <- substr(AMR_genes_NCBI_refseq$refSeqProt, 1, regexpr("\\.", AMR_genes_NCBI_refseq$refSeqProt)-1)
AMR_genes_NCBI_refSeqNuc <- substr(AMR_genes_NCBI_refseq$refSeqNuc, 1, regexpr("\\.", AMR_genes_NCBI_refseq$refSeqNuc)-1)
AMR_genes_NCBI_refseq_clean <- cbind(AMR_genes_NCBI_refSeqProt,AMR_genes_NCBI_refSeqNuc)
class(AMR_genes_NCBI_refseq_clean)
AMR_genes_NCBI_refseq_clean <- as.data.frame(AMR_genes_NCBI_refseq_clean)

```

```{r quantify_data, include=TRUE}

# annotate the scoary output with genbank ids 



sigGenes <- sapply(dataFiles,FUN = function(x) subset(x, Empirical_p<0.05 ),simplify = FALSE,USE.NAMES = TRUE)
sigResistAssoc <- sapply(sigGenes,FUN = function(x) subset(x, Odds_ratio>1 ),simplify = FALSE,USE.NAMES = TRUE)

# also filter by bonferroni p 
sigResistAssocBonf <- sapply(sigResistAssoc,FUN = function(x) subset(x, Bonferroni_p<0.05),simplify = FALSE,USE.NAMES = TRUE)

# want to do an anti-join from dplyr 
# this gives all rows from x where there are not matching values in y 
# x = dataFiles y=AMR_genes_NCBI 
# this will give us the novel AMR-associated genes 

amikacin <- sigResistAssoc$`./Amikacin.INT_15_04_2021_1732.results.csv`
amox_clauv_acid <- sigResistAssoc$`./Amoxicillin.Clavulanic.Acid.INT_15_04_2021_1732.results.csv`
# amox <- sigResistAssoc$`./Amoxicillin.INT_15_04_2021_1732.results.csv`
amp <- sigResistAssoc$`./Ampicillin.INT_15_04_2021_1732.results.csv`
cefalexin <- sigResistAssoc$`./Cefalexin.INT_15_04_2021_1732.results.csv`
#cefalexin and cephalexin are the SAME THING, combine?
cephalexin <- sigResistAssoc$`./Cephalexin.INT_15_04_2021_1732.results.csv`
cefalexin.comb <- rbind(cefalexin,cephalexin)
cefazolin <- sigResistAssoc$`./Cefazolin.INT_15_04_2021_1732.results.csv`
cefovecin <-  sigResistAssoc$`./Cefovecin.INT_15_04_2021_1732.results.csv`
# cefoxitin <- sigResistAssoc$`./Cefoxitin.INT_15_04_2021_1732.results.csv`
cefpodoxime <- sigResistAssoc$`./Cefpodoxime.INT_15_04_2021_1732.results.csv`
ceftazidime <- sigResistAssoc$`./Ceftazidime.INT_15_04_2021_1732.results.csv`
# ceftiofur <- sigResistAssoc$`./Ceftiofur.INT_15_04_2021_1732.results.csv`
# cephalothin <- sigResistAssoc$`./Cephalothin.INT_15_04_2021_1732.results.csv`
chloram <- sigResistAssoc$`./Chloramphenicol.INT_15_04_2021_1732.results.csv`
# clindamycin <- sigResistAssoc$`./Clindamycin.INT_15_04_2021_1732.results.csv`
doxy <- sigResistAssoc$`./Doxycycline.INT_15_04_2021_1732.results.csv`
enro <- sigResistAssoc$`./Enrofloxacin.INT_15_04_2021_1732.results.csv`
# erythro <- sigResistAssoc$`./Erythromycin.INT_15_04_2021_1732.results.csv`
gentamicin <- sigResistAssoc$`./Gentamicin.INT_15_04_2021_1732.results.csv`
imipenem <- sigResistAssoc$`./Imipenem.INT_15_04_2021_1732.results.csv`
marbo <- sigResistAssoc$`./Marbofloxacin.INT_15_04_2021_1732.results.csv`
# neo <- sigResistAssoc$`./Neomycin.INT_15_04_2021_1732.results.csv`
# nitro <- sigResistAssoc$`./Nitrofurantoin.INT_15_04_2021_1732.results.csv`
orbi <- sigResistAssoc$`./Orbifloxacin.INT_15_04_2021_1732.results.csv`
# oxacillin <- sigResistAssoc$`./Oxacillin...2..NaCl.INT_15_04_2021_1732.results.csv`
# penicillin.G <- sigResistAssoc$`./Penicillin.G.INT_15_04_2021_1732.results.csv`
# penicillin <- sigResistAssoc$`./Penicillin.INT_15_04_2021_1732.results.csv`
# piperacillin <- sigResistAssoc$`./Piperacillin.INT_15_04_2021_1732.results.csv`
pipTazo <- sigResistAssoc$`./Piperacillin.Tazobactam.INT_15_04_2021_1732.results.csv`
prado <- sigResistAssoc$`./Pradofloxacin.INT_15_04_2021_1732.results.csv`
tetracycline <- sigResistAssoc$`./Tetracycline.INT_15_04_2021_1732.results.csv`
# ticarc_clauv_acid <- sigResistAssoc$`./Ticarcillin.Clavulanic.Acid.INT_15_04_2021_1732.results.csv`
# ticarcillin <- sigResistAssoc$`./Ticarcillin.INT_15_04_2021_1732.results.csv`
# tobramycin <- sigResistAssoc$`./Tobramycin.INT_15_04_2021_1732.results.csv`
trimeth_sulfameth <- sigResistAssoc$`./Trimethoprim.Sulfamethoxazole.INT_15_04_2021_1732.results.csv`



# first need to combine our scoary results with the gene_key
amikacin_anno <- inner_join(gene_key,amikacin,by=c("qseqid"="Gene"))
# what is missing?
amikacin_anno.missing <- anti_join(amikacin,amikacin_anno,by=c("Gene"="qseqid"))
# discrepancy is genes that are core genes not accessory - these are omitted from the gene_key and therefore do not show up in the output here
# did we find any novel genes?
# probelm: case sensitivity
# amikacin.low <- amikacin_anno %>% mutate(GeneID = tolower(GeneID))
# let's try this again
# anti-join: are there any novel genes
# amikacin_novel <- anti_join(amikacin.low, AMR_genes_NCBI.low,by=c("GeneID" = "gene.family"))
# inner-join: are there any known genes
# amikacin_known <- inner_join(amikacin.low, AMR_genes_NCBI.low,by=c("GeneID" = "gene.family"))


# n_genes <- function(x) {
#  x.anno <- inner_join(x,gene_key,by=c("Gene"="Query"))
#  x.low <- x.anno %>% mutate(GeneID = tolower(GeneID))
#  x.novel <- anti_join(x.low, AMR_genes_NCBI.low,by=c("GeneID" = "gene.family"))
#  x.known <- inner_join(x.low, AMR_genes_NCBI.low,by=c("GeneID" = "gene.family"))
#  n.novel <- nrow(x.novel)
#  n.known <- nrow(x.known)
#  n <- list(x.novel,x.known)
#  return(n)
# }
# 
# test.fun <- n_genes(sigResistAssoc$`./Amikacin.INT_15_04_2021_1732.results.csv`)

#loop over dataSig 



```


```{r accGenomeAnalysis,include=TRUE}

get_prot_seq_hm <- function(path_to_gene_data, x, acc_only=T){
  ## inputs:
  #     path_to_gene_data = "path/gene_data.csv"
  #     path_to_gene_pres_abs = "path/gene_presence_absence_roary.csv"
  #     acc_only = T if you want accessory genes sequences, F if you want all gene sequences 
  
  #read in gene data (contains sequences) and gene presence absence matrix
  genedf <- read.csv(path_to_gene_data)
  genepa <- x
  n = genepa[1,5]
  #only retain distinct protein sequences (only want one sequence per gene)
  genedf_f <- genedf %>% distinct(prot_sequence,.keep_all=T)
  
  #get gene presence absence matrix in long format
  genepa_l <- suppressWarnings(genepa %>% gather(key = "seqid", val = "annotation_id",-colnames(genepa[1:14]))) 
  
  #combine two dataframes
  genepa_comb <- left_join(genedf_f, genepa_l, by="annotation_id")
  
  #only keep one protein sequence per gene
  genepa_comb <- genepa_comb %>% distinct(Gene, .keep_all=T)
  
  
  if (acc_only == T){
    #keep accessory genes
    genepa_comb <- genepa_comb %>% filter(No..isolates <= n*.95)
  }
  
  return(genepa_comb)
} 

write_fasta_out_nt <- function(genepa_comb, path_out){
  # write out protein fasta
  sink(path_out)
  for (i in 1:nrow(genepa_comb)){
    name = noquote(paste(">",genepa_comb[i,9],sep=""))
    cat(name,sep="\n")
    ln <- noquote(as.character(genepa_comb[i,6]))
    cat(ln)
    cat("\n")
  }
  sink()
}

get_nt_fasta <- function(antibiotic,name) {
 x <- inner_join(gene_pres_abs, antibiotic,by=c("Gene" = "Gene"))
 y <- get_prot_seq_hm("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/gene_data.csv",x,T)
write_fasta_out_nt(y, paste("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/",name,".fasta",sep=""))
} 


# get_nt_fasta(sigResistAssocBonf$`./Amoxicillin.Clavulanic.Acid.INT_15_04_2021_1732.results.csv`,"clavamox")


# Want to get a fasta file of the sequences from the CARD output 
# Need to query these produced fasta files with the contigs from CARD 

```

```{r data_wrangling, include=TRUE}
# load in AMRFinder results combined 
amrResultsAll <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMRFinder_results/AMRFinderResultsCombined.csv",header=TRUE)
nlevels(as.factor(amrResultsAll$Isolate))
#942 levels here 

# need to correlate the run (SRR******) with the Assembly (GCA****) naming conventions 
anno <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)

# want Run (ID used in AMRFinder data) and assembly (ID used in Scoary data)
header <- colnames(anno)
header.want <- c(grep("Run",header),grep("Assembly",header))
anno.want <- anno[header.want]


# need to extract the lines from AMRFinder data that match what isolates we have (the 601 dog genomes that are definitely dogs and we have phenotypic data for)

# get the gene presence-absence matrix in long format first
genePA_long <- suppressWarnings(gene_pres_abs %>% gather(key = "seqid", val = "annotation_id",-colnames(gene_pres_abs[1:14]))) 
#seqid is the one we want in this one (GCA*****.1)
seqid <- levels(as.factor(genePA_long$seqid))
seqid <- as.data.frame(seqid)
# combine that with the annotation file so we can get out the 601 seqs we have
anno.want2 <- inner_join(anno.want, seqid, by=c("Assembly"="seqid"))
  
# get the 601 sequences from amrResultsAll
amrResults601 <- inner_join(amrResultsAll,anno.want2,by=c("Isolate"="Run"))
nlevels(as.factor(amrResults601$Isolate))
# 554 levels here 
# what am I missing?
missingSamples <- anti_join(anno.want2,amrResultsAll,by=c("Run"="Isolate"))
#output these to a file so I can download them onto the cluster and run AMRFinder on them 
write.csv(missingSamples$Run,file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/missingfrom601.csv")


#Ran AMRFinder, now need to read in the output for the missing samples 
rest_of_samples <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ARG_table_20210519_160358.csv")
#combine with the original list of samples 
dog_samples <- rbind(amrResultsAll, rest_of_samples)
# get out the ones we want 
dog_samples_601 <- inner_join(dog_samples,anno.want2,by=c("Isolate"="Run"))
#make sure it has the correct number of levels 
nlevels(as.factor(dog_samples_601$Isolate))
# how am I still missing one??
missingStill <- anti_join(anno.want2,dog_samples_601,by=c("Run"="Isolate"))
# this one didn't have a "Run" column filled in 
# did some NCBI digging and found it 
# details on Github issue 

# read in the new sample's ARG table 
last_sample <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ARG_table_20210520_103455.csv")

#combine them all together finally 
all_dogs <- rbind(dog_samples_601,last_sample)
#make sure it has the correct number of levels 
nlevels(as.factor(all_dogs$Isolate))
# success! 


```

```{r AMRFinderCorr, include=TRUE}
# Want to match our Scoary outputs to the AMRFinder outputs and find what matches and what doesn't match, for each Antibiotic tested

all_dogs_AMRFinder <- all_dogs
#make all gene names lowercase for ease of comparison 
all_dogs_AMRFinder.low <- all_dogs_AMRFinder %>% mutate(Gene = tolower(Gene))
amikacin.low <- amikacin %>% mutate(Gene= tolower(Gene))
# inner join with Gene = GeneID
amk_amrF <- inner_join(amikacin.low,all_dogs_AMRFinder.low,by=c("Gene"="Gene"))
clavamox_amrF <- inner_join(amox_clauv_acid,all_dogs_AMRFinder.low,by=c("Gene"="Gene"))

x=amikacin
# # n_genes <- function(x) {
#  x.anno <- inner_join(x,key.low,by=c("Gene"="qseqid"))
#  x.missing <- anti_join(x,key.low,by=c("Gene"="Query"))
#  x.low <- x.anno %>% mutate(GeneID = tolower(GeneID))
#  # x.novel <- anti_join(x.low, all_dogs_AMRFinder.low,by=c("GeneID" = "Gene"))
#  known.1 <- grepl(all_dogs_AMRFinder.low$Gene, x.low$Gene)
#  known.2 <- grepl(all_dogs_AMRFinder.low, x$Query)
#  # n.novel <- nrow(x.novel)
#  n.known <- nrow(x.known)
#  # n <- list(x.novel,x.known)
#  return(n.known)
# }
# 
# # test.fun <- n_genes(tetracycline)


```

```{r create itol files, include=TRUE}

#Script from Kristina Ceres

# Make into a function 

## read in scoary results and gene presence absence matrix
# (we already have these)
# setwd("~/PhD/ecoli_host_adaptation/dog_confirmed_host/")
# genepa <- read.csv("gene_presence_absence_roary.csv")
# clavamox <- read.csv("Amoxicillin.Clavulanic.Acid.INT_15_04_2021_1732.results.csv")

setwd("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/")
abx=amp
fileName="amikacin"
phenoName="Amikacin.INT"
create_itol_files <- function(abx,fileName,phenoName) {
  # select significant results
abx_filt <- abx %>% filter(Empirical_p < 0.05 & Bonferroni_p < 0.05) 
abx_filt$log_or <- log10(abx_filt$Odds_ratio)
abx_filt_l <- c(abx_filt$Gene)

# filter gene presence absence matrix to keep only genes significanctly
# associated with clavamox resistance
genepa_abx_filt <- gene_pres_abs %>% filter(Gene %in% abx_filt_l)
genepa_abx_filt_combined <- left_join(genepa_abx_filt, abx_filt)
genepa_filtered <- genepa_abx_filt[,15:ncol(genepa_abx_filt)]
rownames(genepa_filtered) = genepa_abx_filt$Gene

# create matrix to store future ITOL table
gene_mat <- matrix(NA, nrow=nrow(genepa_filtered), ncol=ncol(genepa_filtered))

# populate matrix with log odds ratio if gene is present, 
# -2 if gene is absent (-2 chosen to aid in coloring on ITOL scale)
for (r in 1:nrow(genepa_filtered)){
  for (c in 1:ncol(genepa_filtered)){
    if (genepa_filtered[r,c] != ""){
      gene_mat[r,c]= genepa_abx_filt_combined$log_or[r] 
    } else{
      gene_mat[r,c]= -2#0
    }
  }
}

colnames(gene_mat) = colnames(genepa_filtered)
rownames(gene_mat) = rownames(genepa_filtered)

gene_mat_itol <- t(gene_mat)

# write out table for ITOL 
write.table(gene_mat_itol,paste("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/Scoary_Results/gene_mat_itol",fileName,".txt",sep=""), quote=F)

## get phenotype data in itol format 
pheno <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/dog_verified_host_PhenoForScoary.csv")
pheno <- pheno %>% dplyr::select(X,all_of(phenoName))
write.table(pheno, paste("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/pheno_4itol",fileName,".txt",sep=""),quote=F, row.names = F)
}

create_itol_files(trimeth_sulfameth,"trimeth_sulfameth","Trimethoprim.Sulfamethoxazole.INT")

 



create_itol_files_emp_p <- function(abx,fileName,phenoName) {
  # select significant results
abx_filt <- abx %>% filter(Empirical_p < 0.05 & Bonferroni_p < 0.05) 
abx_filt$log_or <- log10(abx_filt$Odds_ratio)
abx_filt_l <- c(abx_filt$Gene)

# filter gene presence absence matrix to keep only genes significanctly
# associated with clavamox resistance
genepa_abx_filt <- gene_pres_abs %>% filter(Gene %in% abx_filt_l)
genepa_abx_filt_combined <- left_join(genepa_abx_filt, abx_filt)
genepa_filtered <- genepa_abx_filt[,15:ncol(genepa_abx_filt)]
rownames(genepa_filtered) = genepa_abx_filt$Gene

# create matrix to store future ITOL table
gene_mat <- matrix(NA, nrow=nrow(genepa_filtered), ncol=ncol(genepa_filtered))

# populate matrix with log odds ratio if gene is present, 
# -2 if gene is absent (-2 chosen to aid in coloring on ITOL scale)
for (r in 1:nrow(genepa_filtered)){
  for (c in 1:ncol(genepa_filtered)){
    if (genepa_filtered[r,c] != ""){
      gene_mat[r,c]= genepa_abx_filt_combined$Empirical_p[r] 
    } else{
      gene_mat[r,c]= 0#0
    }
  }
}

colnames(gene_mat) = colnames(genepa_filtered)
rownames(gene_mat) = rownames(genepa_filtered)

gene_mat_itol <- t(gene_mat)

# write out table for ITOL 
write.table(gene_mat_itol,paste("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/Scoary_Results/gene_mat_itol",fileName,".txt",sep=""), quote=F)

## get phenotype data in itol format 
pheno <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/dog_verified_host_PhenoForScoary.csv")
pheno <- pheno %>% dplyr::select(X,all_of(phenoName))
write.table(pheno, paste("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/pheno_4itol",fileName,".txt",sep=""),quote=F, row.names = F)
}

create_itol_files_emp_p(prado,"prado_emp_p","Pradofloxacin.INT")

pradoGenes <- prado$Gene
write.table(prado,"/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/prado.txt")
prado.fixed <- read.table("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/prado.txt",header=TRUE)
sigGenesPrado <- pradoGenes %in% gene_key$qseqid


```

```{r ggTree, include=TRUE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("treeio")
library(treeio)

dogTree <- read.tree("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/dog_verified_host_gb_snp_sites.fasta.treefile")

dogTreePhy <- as.phylo(dogTree)

library(ape)
x <- as_tibble(dogTree)
x


pheno <- read.csv("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/dog_verified_host_PhenoForScoary.csv")

y <- full_join(x, pheno, by =c("label"="X"))
y

as.treedata(y)

```

```{r scoaryPlots,include=TRUE}
# from https://github.com/cimendes/scoaryPlots

if (!require("hexbin")){
  install.packages("hexbin",dependencies = TRUE)
  library('hexbin')
}

#read chosen file
file = "/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Scoary_Results/Ampicillin.INT_15_04_2021_1732.results.csv"
filename = basename(file)

data=read.csv(file, header=TRUE, sep=',')

#select data by benjamin corrected p-value
data.df=data.frame(data)
#names(data.df)
data.df.sub=subset(data.df,Benjamini_H_p<0.05)

# ****** I am making my own script here ********* # 
# I want to know how many genes are significantly correlated with resistance (positive PT)
 # I also want to know WHAT those genes are and what they DO (GO analysis - will use another script later for this)
  # First step: filter by empircal p value (this takes into account the tree) 
  data.df.sub=subset(data.df,Empirical_p<0.05) 
    # Next: extract columns interested in - # posititves present in, sensitivity, and specificity 
    data.df.sub2 <- data.df.sub[, c("Gene", "Non.unique.Gene.name","Annotation","Number_pos_present_in","Sensitivity","Specificity","Empirical_p")]
# how many genes are significantly correlated with resistance phenotype?
    # plot number positive present in by Empirical_p? 
    ggplot(data=data.df.sub2, mapping = aes( data.df.sub2$Empirical_p, data.df.sub2$Number_pos_present_in,col=data.df.sub2$Specificity) ) + geom_point()
    
    
    
    
 ## ******* end of my own script ************ ## 
    
    
    
    
    
#Select data
positive_present_in=data.df.sub[,4]
negative_present_in=data.df.sub[,5]

#save plot
png(paste(file,"_scatterplot_gray.png"), width=480, height=480)

#create plot

cols <- colorRampPalette(c("darkblue", "red"))
bANDw <- colorRampPalette(c("lightgray","black"))

#plot - gray scale
bin<-hexbin(negative_present_in,positive_present_in)
#cols <- colorRampPalette(c("lightgray","black"))
plot(bin, main=filename,colorcut = seq(0,1,length.out=15),
     colramp = function(n) bANDw(15))
dev.off()

#plot - colours
png(paste(file,"_scatterplot_colour.png"), width=480, height=480)
#cols <- colorRampPalette(c("darkblue","red"))
plot(hexbin(negative_present_in,positive_present_in), main = filename ,
     colorcut = seq(0,1,length.out=15),
     colramp = function(n) cols(15))
dev.off()

#plot - control - all data
positive_present_in_alldata=data.df[,4]
negative_present_in_alldata=data.df[,5]
png(paste(file,"_alldata.png"), width=480, height=480)
#cols <- colorRampPalette(c("darkblue","red"))
plot(hexbin(negative_present_in_alldata,positive_present_in_alldata), main = filename ,
     colorcut = seq(0,1,length.out=15),
     colramp = function(n) cols(15))
dev.off()


#Exclusive accessory genome (10% range)
max.Number_pos_present_in=max(positive_present_in)
range.Number_pos_present_in=max.Number_pos_present_in-max.Number_pos_present_in*0.1
subset.exclusive=subset(data.df,Number_pos_present_in>=range.Number_pos_present_in & Number_neg_present_in == 0)
if (length(row.names(subset.exclusive)) != 0){
  write.table(subset.exclusive, file=paste(file,"exclusive_present_range.csv"), sep=";", row.names = FALSE)
  } else{
  print(paste("No value to print in ", file, "exclusive present"))
  }
max.Number_neg_present_in=max(negative_present_in)
range.Number_neg_present_in=max.Number_neg_present_in-max.Number_neg_present_in*0.1
subset.absent=subset(data.df,Number_pos_present_in==0 & max.Number_neg_present_in >= range.Number_neg_present_in)
if (length(row.names(subset.absent)) != 0){
  write.table(subset.absent, file=paste(file,"exclusive_absent_range.csv"), sep=";", row.names = FALSE)
  } else{
  
  print(paste("No value to print in ", file, "exclusive absent"))
  }




```

```{r GO analysis, include=TRUE}
# Get gene names from each abx 
# Use gene key and inner_join 
# Try with the sapply/lapply function
class(sigResistAssoc$`./Amikacin.INT_15_04_2021_1732.results.csv`$Gene)
class(key.low$Query)
genekeyRn <- key.low
colnames(genekeyRn) <- c("Gene","ID")
sigResistAssoc_genes <- sapply(sigResistAssoc_chr,FUN = function(x) inner_join(x,genekeyRn,by=character(),copy=TRUE),simplify = FALSE,USE.NAMES = TRUE)


tog <- inner_join(amikacin,genekeyRn,by=c("Gene"="Gene"))
abxGenes <- as.data.frame(tog$ID)
write.table(abxGenes, file=paste("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/",".txt"))


get_genes <- function(abx,name) {
tog <- inner_join(abx,genekeyRn,by=c("Gene"="Gene"))
abxGenes <- as.data.frame(tog$ID)
write.table(abxGenes, file=paste("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/Holly_Analyses/",name,".txt",sep=""))
}

get_genes(amikacin,"amikacin")



```

```{r heatmap, include=TRUE,echo=FALSE}
# get giant matrix of all sig (emp p < 0.05) assoc genes for all abx
# annotate each abx with its name 
amikacin["Abx"] <- "amikacin"
amp["Abx"] <- "ampicillin"
amox_clauv_acid["Abx"] <- "amox_clauv_acid"
cefalexin.comb["Abx"] <- "cefalexin"
cefazolin["Abx"] <- "cefazolin"
cefovecin["Abx"] <- "cefovecin"
cefpodoxime["Abx"] <- "cefpodoxime"
ceftazidime["Abx"] <- "ceftazidime"
chloram["Abx"] <- "chloramphenicol"
doxy["Abx"] <- "doxycycline"
enro["Abx"] <- "enrofloxacin"
gentamicin["Abx"] <- "gentamicin"
imipenem["Abx"] <- "imipenem"
marbo["Abx"] <- "marbofloxacin"
orbi["Abx"] <- "orbifloxacin"
pipTazo["Abx"] <- "pipTazo"
prado["Abx"] <- "prado"
tetracycline["Abx"] <- "tetracycline"
trimeth_sulfameth["Abx"] <- "trimeth_sulfameth"

giantMatrixAllAbxSig <- rbind(amikacin,amp,amox_clauv_acid,cefalexin.comb,cefazolin,cefovecin, cefpodoxime,ceftazidime, chloram,doxy,enro,gentamicin,imipenem,marbo,orbi,pipTazo, prado, tetracycline,trimeth_sulfameth)
nlevels(as.factor(giantMatrixAllAbxSig$Abx))
# further break it down to bonferonni p < 0.05 
giantMatrixAllAbxSigEmpBonf <- subset(giantMatrixAllAbxSig, Bonferroni_p<0.05) #, simplify = FALSE,USE.NAMES = TRUE)
nlevels(as.factor(giantMatrixAllAbxSigEmpBonf$Abx))
#3 abx that had 0 genes with bonf < 0.05 (these were: amikacin, impipenem, and pipTazo)
levels(as.factor(giantMatrixAllAbxSigEmpBonf$Abx))
# make new column with the number of isolates the gene is found in 
head(as.numeric(giantMatrixAllAbxSigEmpBonf$Number_pos_present_in))
giantMatrixAllAbxSigEmpBonf$Number_pos_present_in <- as.numeric(giantMatrixAllAbxSigEmpBonf$Number_pos_present_in )
giantMatrixAllAbxSigEmpBonf$Number_neg_present_in <- as.numeric(giantMatrixAllAbxSigEmpBonf$Number_neg_present_in)
giantMatrixAllAbxSigEmpBonf$Number_present_in <- giantMatrixAllAbxSigEmpBonf$Number_pos_present_in + giantMatrixAllAbxSigEmpBonf$Number_neg_present_in
# extract out only the columns we're interested in (can always get the others later)
giantMatrixAllAbxSigEmpBonfSml <- giantMatrixAllAbxSigEmpBonf[,c(1,3,10,12,19,20,21)]
as.factor(giantMatrixAllAbxSigEmpBonfSml$Gene)
nlevels(as.factor(giantMatrixAllAbxSigEmpBonfSml$Gene))
nlevels(as.factor(giantMatrixAllAbxSigEmpBonfSml$Abx))

# want a new column with the number of times that gene is seen in the entire dataset
# What I actually wanted: # of isolates that gene was found in 
# What I got instead: # of antibiotics that gene was found in 
library(dplyr)
giantMatrixAllAbxSigEmpBonfSmlCount <- giantMatrixAllAbxSigEmpBonfSml %>% group_by(Gene) %>% mutate(Num_Abx_Pres_In=n_distinct(Abx))
colnames(giantMatrixAllAbxSigEmpBonfSmlCount)
# yay! This is exactly what I want! 

library(ggplot2)
library(viridis)
heatmap <- ggplot(giantMatrixAllAbxSigEmpBonfSmlCount, aes(x = Abx, Gene)) +
  geom_tile(aes(fill = Number_present_in)) + theme_classic() + theme(axis.text.y = element_text(size = 2),axis.text.x = element_text(angle=45,vjust=0.5,size = 8))  + scale_fill_viridis(option="magma",direction=-1)


pdf("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/Visualizations/heatmap_num_pres_in.pdf")
heatmap
dev.off()



# Okay let's try pheatmap now, because it allows us to cluster things and also makes way prettier heatmaps :) 
library(pheatmap)
 
# need to convert data into a numerical matrix 
# with abx as column names, and genes as row names, counts as the numerical values
#get just these columns out 

# try hierarchical clustering manually first 
# using Odds ratio and/or empirical p value 
str(giantMatrixAllAbxSigEmpBonfSmlCount)
# clusters <- hclust(dist(giantMatrixAllAbxSigEmpBonfSmlCount[, 1,3]))

ecoliMatrix <- giantMatrixAllAbxSigEmpBonfSmlCount[,c(1,2,3,6,7,8)]
typeof(ecoliMatrix)
str(ecoliMatrix)
typeof(ecoliMatrix$Num_Abx_Pres_In)
ecoliMatrix$Num_Abx_Pres_In <- as.numeric(ecoliMatrix$Num_Abx_Pres_In)
ecoliMatrix$Number_present_in <- as.numeric(ecoliMatrix$Number_present_in)
typeof(ecoliMatrix$Num_Abx_Pres_In)
nlevels(as.factor(ecoliMatrix$Abx))

library(tidyr)
# The arguments to spread():
# - data: Data object
# - key: Name of column containing the new column names
# - value: Name of column containing values
library(reshape2)

data_wide_num_abx <- dcast(ecoliMatrix, Gene + Annotation ~ Abx, value.var="Num_Abx_Pres_In")
data_wide_num_abx
write.csv(data_wide_num_abx,file="/Users/hcm59/Box/Holly/Dog_E_coli_project/data_wide.csv")
rownames(data_wide_num_abx) <- data_wide_num_abx$Gene
heatmap_data_num_abx <- data_wide_num_abx[,c(3:18)]
str(heatmap_data_num_abx)
typeof(heatmap_data_num_abx)
heatmap_data_num_abx <- as.data.frame(heatmap_data_num_abx)
typeof(heatmap_data_num_abx)
sigGenesAll <- heatmap_data_num_abx

data_wide_num_iso <- dcast(ecoliMatrix, Gene + Annotation ~ Abx, value.var="Number_present_in")
data_wide_num_iso
write.csv(data_wide_num_iso,file="/Users/hcm59/Box/Holly/Dog_E_coli_project/data_wide.csv")
rownames(data_wide_num_iso) <- data_wide_num_iso$Gene
heatmap_data_num_iso <- data_wide_num_iso[,c(3:18)]
str(heatmap_data_num_iso)
typeof(heatmap_data_num_iso)
heatmap_data_num_iso <- as.data.frame(heatmap_data_num_iso)
typeof(heatmap_data_num_iso)
sigGenesAll <- heatmap_data_num_iso

# sigGenesAll <- cbind(rownames(sigGenesAll), data.frame(sigGenesAll, row.names=NULL))
# sigGenesAllAnno <- inner_join(sigGenesAll,gene_pres_abs,by=c("rownames(sigGenesAll)"="Gene"))
# sigGenesAllAnno2 <- cbind(sigGenesAllAnno[,c(1,15:17,2:14)])
# write.csv(sigGenesAllAnno2,file="/Users/hcm59/Box/Holly/Dog_E_coli_project/sigGenesAllAnno.csv")

# Need to break up sigGenesAllAnno file into antibiotic classes
# Misc: Gentamicin, chloramphenicol, trimeth-sulfameth
misc <- data_wide[,c(10,13,18)]
misc.1 <- misc[rowSums(is.na(misc)) != ncol(misc), ]
write.csv(misc.1,file="miscellaneuous_abx.csv")
# penicillions: ampicillin, clavamox
penicillins <- data_wide[,c(3,4)]
penicillins.1 <- penicillins[rowSums(is.na(penicillins)) != ncol(penicillins), ]
# cephalosporins: cefalexin, cefazolin, cefovecin, cefpodoxime, ceftazidime
cephalosporins <- data_wide[,c(5:9)]
cephalosporins.1 <- cephalosporins[rowSums(is.na(cephalosporins)) != ncol(cephalosporins), ]
# fluoroquinolones: enrofloxacin, marbofloxacin, orbifloxacin,pradofloxacin 
fluoroquinolones <- data_wide[,c(12,14:16)]
fluoroquinolones.1 <- fluoroquinolones[rowSums(is.na(fluoroquinolones)) != ncol(fluoroquinolones), ]
# tetracyclines: doxycycline, tetracyclineb
tetracyclines <- data_wide[,c(11,17)]
tetracyclines.1 <- tetracyclines[rowSums(is.na(tetracyclines)) != ncol(tetracyclines), ]

pheatmap(heatmap_data,main = "pheatmap default",cluster_rows = F, cluster_cols = F)


pdf("misc_heatmap.pdf")
pheatmap(misc.1, main = "misc",cluster_rows = F, cluster_cols = F,fontsize_row=2.5)
dev.off()

pdf("penicillins_heatmap.pdf")
pheatmap(penicillins.1, main = "Penicillins",cluster_rows = F, cluster_cols = F,fontsize_row=2.5)
dev.off()

pdf("cephalosporins_heatmap.pdf")
pheatmap(cephalosporins.1, main = "Cephalosporins",cluster_rows = F, cluster_cols = F,fontsize_row=2.5)
dev.off()

pdf("fluoroquinolones_heatmap.pdf")
pheatmap(fluoroquinolones.1, main = "Fluroquinolones",cluster_rows = F, cluster_cols = F,fontsize_row=2.5)
dev.off()

pdf("tetracyclines_heatmap.pdf")
pheatmap(tetracyclines.1, main = "Tetracyclines",cluster_rows = F, cluster_cols = F,fontsize_row =2.5)
dev.off()

library(colorspace)
testHeatmap <- pheatmap(tetracyclines.1, 
         color = sequential_hcl(13,palette = "Mako"),
         cluster_rows = FALSE,
         cluster_cols = FALSE,
         # gaps_row = head(as.numeric(cumsum(gaps)),-1), 
         # cutree_cols = 3,
         fontsize = 2,
         fontsize_col = 8,
         show_rownames = TRUE
         )
pdf("test.pdf")
testHeatmap
dev.off()

```



## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
