---
title: "treeWAS"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# install packages
library(treeWAS)

```

```{r load_pheno_data, include=FALSE}
# load in excel file of phenotype data 
phen2017 <- read.csv(file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/2017-ECOL-MIC-Vet-LIRN.csv",header=TRUE,na.strings = c("", "NA"))
# View(phen2017)
# need to subset data into what I actually need 
# what I need:
# sample name (the one that matches the sequencing data)
# antibiotic name 
  # interpretation of susceptibility/resistance to specified antibiotic 
# headr <- colnames(phen2017)
# headr.want <- c(grep("sample.ID",headr),grep("INT",headr))
# headr.want 
# phen2017.sub <- phen2017[headr.want]
# View(phen2017.sub)

### do the same for 2018 and 2019 datasets
# load in excel file of phenotype data 
phen2018 <- read.csv(file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/2018-ECOL-MIC-Vet-LIRN.csv",header=TRUE,na.strings = c("", "NA"))
# View(phen2018)
# need to subset data into what I actually need 
# what I need:
# sample name (the one that matches the sequencing data)
# antibiotic name 
  # interpretation of susceptibility/resistance to specified antibiotic 
# headr <- colnames(phen2018)
# headr.want <- c(grep("sample.ID",headr),grep("INT",headr))
# headr.want 
# phen2018.sub <- phen2018[headr.want]
# View(phen2018.sub)

### do the same for 2018 and 2019 datasets
# load in excel file of phenotype data 
phen2019 <- read.csv(file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/2019-ECOL-MIC-Vet-LIRN.csv",header=TRUE,na.strings = c("", "NA"))
# # View(phen2019)
# # need to subset data into what I actually need 
# # what I need:
# # sample name (the one that matches the sequencing data)
# # antibiotic name 
#   # interpretation of susceptibility/resistance to specified antibiotic 
# # headr <- colnames(phen2019)
# # headr.want <- c(grep("sample.ID",headr),grep("INT",headr))
# # headr.want 
# # phen2019.sub <- phen2019[headr.want]
# # View(phen2019.sub)
# # 
# # dim(phen2019.sub)
# # 
# # colnames(phen2019.sub) == colnames(phen2017.sub)
# # colnames(phen2019.sub) == colnames(phen2018.sub)
# # colnames(phen2018.sub) == colnames(phen2017.sub)
# 
# # use reshape2 to melt data (want 3 cols: sample.ID, antibiotic, INT)
#   # use melt data frame (https://www.rdocumentation.org/packages/reshape2/versions/1.4.4/topics/melt.data.frame)
#     # don't include any value for measure.vars so that it uses everything except sample.ID.
# # phen2017.melt <- melt(phen2017.sub, id.vars=c("sample.ID."),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)
# # View(phen2017.melt)
# # phen2018.melt <- melt(phen2018.sub, id.vars=c("sample.ID."),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)
# # View(phen2018.melt)
# # phen2019.melt <- melt(phen2019.sub, id.vars=c("sample.ID."),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)
# # View(phen2019.melt)
# # 
# # str(phen2017.melt)
# # str(phen2018.melt)
# # str(phen2019.melt)
# 
# 
# # stick all three years together
# # phenAll <- rbind(phen2017.melt,phen2018.melt,phen2019.melt)
# # View(phenAll)


### MIC Cutoffs from CLSI VET01SEd5E performance standards document
# Amikacin S <= 4  R > 4
# cutoffs <- 




```


```{r clean_up_data,include=FALSE}
# want to know what example data looks like
# data(snps)
# View(snps)
# data(phen)
# View(phen)
# data(tree)
# View(tree)
# names(phen)
# str(phen)
# typeof(phen)
# out <- treeWAS(snps = snps,
#                 phen = phen,
#                 tree = tree,
#                 seed = 1)
# # 
# names(phen)
# typeof(phen)
# str(phen)
# all(rownames(snps) %in% names(phen))
# all(names(phen) %in% rownames(snps))
# 
# 
# ## Examine data:
# ## genetic data
# str(snps)

## phenotype
# str(phen)
# table(phen)
# phen <- as.data.frame(phen)
# View(phen)
# 
# ## tree
# str(tree)

## Read data from file:
eColiSNPs <- read.dna(file = "/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/dog_verified_host_gb_snp_sites.fasta", format = "fasta")
rownames(eColiSNPs)
colnames(eColiSNPs)
str(eColiSNPs)
## Convert: 
eColiMat <- DNAbin2genind(eColiSNPs)@tab

# head(eColiMat)
rownames(eColiMat)
# colnames(eColiMat)

## annotations file that contains IDs that match phenotype dataset and those that match SNP dataset 
anno <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)




# want strain (ID used in phenotypic data) and assembly (ID used in SNP data)
header <- colnames(anno)
header.want <- c(grep("Strain",header),grep("Assembly",header))
anno.want <- anno[header.want]
View(anno.want)
anno.want2 <- substr(anno.want$Assembly, 1, regexpr("\\.", anno.want$Assembly)-1)
anno.want2
eColirnames <- rownames(eColiMat)
eColirnames
eColirnames.sub <- substr(eColirnames, 1, regexpr("\\.", eColirnames)-1)
eColirnames.sub
samples.want <- anno.want2 %in% eColirnames.sub
samples.want
table(samples.want)
anno.want2 <- anno.want[samples.want,]
anno.want2
anno.want <- anno.want2




# merge phenotypic data with annotations to get IDs and assembly in same file 
library(dplyr)
colnames(phen2017)
phen2017.anno <- inner_join(anno.want, phen2017, by=c("Strain" = "sample.ID."),keep=TRUE)
View(phen2017.anno)
headr2017 <- colnames(phen2017.anno)
headr2017.want <- c(grep("Assembly",headr2017),grep("sample.ID.",headr2017),grep("INT",headr2017),grep("specific.host",headr2017))
                    # ,grep(".MIC",headr2017)
headr2017.want 
phen2017.anno2 <- phen2017.anno[headr2017.want]
dim(phen2017.anno)
rm(headr2017,headr2017.want)
library(dplyr)    
phen2017.fin <- filter(phen2017.anno, specific.host == "Canis lupus familiaris")

phen2018.anno <- inner_join(anno.want, phen2018, by=c("Strain" = "sample.ID."),keep=TRUE)
View(phen2018.anno)
headr2018 <- colnames(phen2018.anno)
headr2018.want <- c(grep("Assembly",headr2018),grep("sample.ID.",headr2018),grep("INT",headr2018),grep("specific.host",headr2018))
                    # ,grep("MIC",headr2018))
headr2018.want 
phen2018.anno <- phen2018.anno[headr2018.want]
dim(phen2018.anno)
rm(headr2018,headr2018.want)
phen2018.fin <- filter(phen2018.anno, specific.host == "Canis lupus familiaris")

phen2019.anno <- inner_join(anno.want, phen2019, by=c("Strain" = "sample.ID."),keep=TRUE)
View(phen2019.anno)
headr2019 <- colnames(phen2019.anno)
headr2019.want <- c(grep("Assembly",headr2019),grep("sample.ID.",headr2019),grep("INT",headr2019),grep("specific.host",headr2019))
                    # ,grep("MIC",headr2019))
headr2019.want 
phen2019.anno <- phen2019.anno[headr2019.want]
dim(phen2019.anno)
rm(headr2019,headr2019.want)
phen2019.fin <- filter(phen2019.anno, specific.host == "Canis lupus familiaris")

rm(phenAll)
phenAll <- dplyr::bind_rows(phen2017.fin,phen2018.fin,phen2019.fin)
phenAll$Assembly
test <- dplyr::inner_join(phenAll,anno.want,by="Assembly")
test$Assembly <- substr(test$Assembly, 1, regexpr("\\.", test$Assembly)-1)
test$Assembly
phenAllFinal <- test
write.csv(phenAllFinal,"/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/phenotypesDogs171819.csv")

rownames(eColiMat)

# phenAll <- read.table(file="/Users/hcm59/Box/Github/CornellPostdoc/phenAll.txt",header=TRUE)
# library(tidyverse)
# export the assembly numbers to a text file so I can extract those sequences from the fasta file for input into IQtree
samples_wanted <- phenAll$Assembly
write.table(samples_wanted,file="/Users/hcm59/Box/Github/CornellPostdoc/samples_wanted.txt")
# need to parse the SNP dataset to get the same samples as phenotypic dataset 
rownames(eColiMat)
str(eColiMat)
typeof(eColiMat)
dim(eColiMat)

eColirnames <- rownames(eColiMat)
eColirnames
eColirnames.sub <- substr(eColirnames, 1, regexpr("\\.", eColirnames)-1)
str(eColirnames.sub)
typeof(eColirnames.sub)
eColiMat2 <- eColiMat
rownames(eColiMat2) <- eColirnames.sub
eColiMat2$Assembly <- rownames(eColiMat2)
combotest <- dplyr::inner_join(phenAll,eColiMat2,by=c("Assembly","Assembly"))


all(rownames(names(phenAllFinal)) %in% rownames(eColiMat2))
all(rownames(eColiMat2) %in%  rownames(names(phenAllFinal))) 


# write.table(eColiMat2,file="/Users/hcm59/Box/Github/CornellPostdoc/subsetEcoliSNPs.txt")
# write.table(phenAllFinal,file="/Users/hcm59/Box/Github/CornellPostdoc/phenAll.txt")

library(ape)
# data(woodmouse)
# str(woodmouse)
# mouseMat<-as.character(woodmouse)
# dim(mouseMat)
# str(mouseMat)
# subsetMouse<-mouseMat[c('No305','No304','No306'),]
# dim(subsetMouse)

### need to subset fasta file for the samples that I have phenotypic data for 
install.packages("remotes")
remotes::install_github("GuillemSalazar/FastaUtils")

```

```{r phen_subset,include=FALSE}
# first import back in the data you exported before
# Phen <- read.table("/Users/hcm59/Box/Github/CornellPostdoc/phenAll.txt",header=TRUE)
Phen <- phenAllFinal
headr <- colnames(Phen)
headr.want <- c(grep("sample.ID",headr),grep("Assembly",headr), grep("INT",headr))
headr.want
phen.sub <- Phen[headr.want]
head(phen.sub)
names(phen.sub)[2]
typeof(phen.sub[2])
typeof(phen.sub[1])

library(dplyr)
library(reshape2)
# phen.Melt <- melt(Phen, id.vars=c("Assembly"),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)


# find all instances of "SUSC" and replace with 0
library(tidyverse)
phen.sub.0 <- phen.sub %>% mutate_if(is.atomic,funs(str_replace(., "SUSC", "0")))
phen.sub.0
# worked! # now do that to RESIST and convert NOINTP
phen.sub.01 <- phen.sub.0 %>%  mutate_if(is.atomic,funs(str_replace(., "RESIST", "1")))
phen.sub.01.1 <- phen.sub.01 %>%  mutate_if(is.atomic,funs(str_replace(., "NOINTP", "")))
phen.sub.01.2 <- phen.sub.01.1 %>%  mutate_if(is.atomic,funs(str_replace(., "INTER", "")))
View(phen.sub.01.2)
phen.sub.01.3 <- phen.sub.01.2 %>% mutate_all(na_if,"")


rownames(phen.sub.01.3) <- phenAllFinal$Assembly
rownames(phen.sub.01.3)
View(phen.sub.01.3)
phen.sub.01.4 <- phen.sub.01.3[,2:104]
View(phen.sub.01.4)
str(phen.sub.01.4)

# phen.melt <- melt(phen.sub.01.4, id.vars=c("Assembly"),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)
phen.Amikacin <- phen.sub.01.4[,2]
names(phen.Amikacin) <- rownames(phen.sub.01.4)
View(phen.Amikacin)
typeof(phen.Amikacin)

phen.Amikacin2 <- as.vector(unlist(phen.Amikacin))
names(phen.Amikacin2) <- names(phen.Amikacin)
write.table(phen.Amikacin2,file="/Users/hcm59/Box/Github/CornellPostdoc/phen_Amikacin.txt")

```


```{r treeWAS_server_script,include=TRUE}
### Script for running treeWAS on the AHDC server

## First need to load in/install treeWAS
## install devtools, if necessary:
# install.packages("devtools", dep=TRUE)
library(devtools)

## install treeWAS from github:
# install_github("caitiecollins/treeWAS", args=c('--library="/home/hcm59/R/x86_64-pc-linux-gnu-library/3.5"'))
library(treeWAS)
library(adegenet)


## Read data from file:

# eColiSNPs <- read.dna(file = "/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/results_with_long_branch_sequence_removed/Ecoli_dog_core_snps.fasta", format = "fasta")
eColiSNPs <- read.dna(file = "/workdir/hcm59/Ecoli/SNPs/dog_verified_host/snp_sites/dog_verified_host_gb_snp_sites.fasta", format = "fasta")
rownames(eColiSNPs)
colnames(eColiSNPs)
str(eColiSNPs)


## Convert: 
eColiMat <- DNAbin2genind(eColiSNPs)@tab

# Remove columns with more than 50% NAs: 
eColiMatShrt <- eColiMat
eColiMatShrt <- eColiMatShrt[, which(colMeans(!is.na(eColiMatShrt)) > 0.5)]

anno <- read.csv("/local/workdir/hcm59/Ecoli/R_analyses/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)
# anno <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)
header <- colnames(anno)
header.want <- c(grep("Strain",header),grep("Assembly",header))
anno.want <- anno[header.want]
# View(anno.want)

phenAll <- read.csv("/workdir/hcm59/Ecoli/SNPs/dog_verified_host/dog_verified_host_PhenoForScoary_OxPolRm.csv")
# phenAll <- read.table("/Users/hcm59/Box/Github/CornellPostdoc/phenAll.txt",header=TRUE)
eColirnames <- rownames(eColiMatShrt)
eColirnames
eColirnames.sub <- substr(eColirnames, 1, regexpr("\\.", eColirnames)-1)
str(eColirnames.sub)
typeof(eColirnames.sub)
eColiMat2 <- eColiMatShrt
rownames(eColiMat2) <- eColirnames.sub

all(rownames(names(phenAll)) %in% rownames(eColiMat2))

Phen <- phenAll
headr <- colnames(Phen)
headr.want <- c(grep("sample.ID",headr),grep("Assembly",headr), grep("INT",headr))
headr.want
phen.sub <- Phen[headr.want]
head(phen.sub)
names(phen.sub)[2]
typeof(phen.sub[2])
typeof(phen.sub[1])

library(dplyr)
library(reshape2)
# phen.Melt <- melt(Phen, id.vars=c("Assembly"),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)


# find all instances of "SUSC" and replace with 0
library(tidyverse)
phen.sub.0 <- phen.sub %>% mutate_if(is.atomic,funs(str_replace(., "SUSC", "0")))
phen.sub.0
# worked! # now do that to RESIST and convert NOINTP
phen.sub.01 <- phen.sub.0 %>%  mutate_if(is.atomic,funs(str_replace(., "RESIST", "1")))
phen.sub.01.1 <- phen.sub.01 %>%  mutate_if(is.atomic,funs(str_replace(., "NOINTP", "")))
phen.sub.01.2 <- phen.sub.01.1 %>%  mutate_if(is.atomic,funs(str_replace(., "INTER", "")))
phen.sub.01.3 <- phen.sub.01.2 %>% mutate_all(na_if,"")


rownames(phen.sub.01.3) <- phenAll$Assembly
rownames(phen.sub.01.3)
phen.sub.01.4 <- phen.sub.01.3[,2:104]
str(phen.sub.01.4)

# phen.melt <- melt(phen.sub.01.4, id.vars=c("Assembly"),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)
phen.Amikacin <- phen.sub.01.4[,2]
names(phen.Amikacin) <- rownames(phen.sub.01.4)
# View(phen.Amikacin)
typeof(phen.Amikacin)
phen.Amikacin

phen.Amikacin2 <- as.vector(unlist(phen.Amikacin))
names(phen.Amikacin2) <- names(phen.Amikacin)
all(rownames(names(phen.Amikacin2)) %in% rownames(eColiMat2))
all(rownames(eColiMat2) %in% names(phen.Amikacin2))
## Need files: phenotypic file, SNP file, tree file
# phen_Ami <- read.table("/local/workdir/hcm59/Ecoli/R_analyses/phen_Amikacin.txt")
# phen.Ami <- as.vector(unlist(phen_Ami))
# names(phen.Ami) <- rownames(phen_Ami)
SNPs <- eColiMat2
Tree <- read.tree("/workdir/hcm59/Ecoli/SNPs/dog_verified_host/dog_verified_host_gb_snp_sites.fasta.treefile")
# Tree <- read.tree("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/results_with_long_branch_sequence_removed/Ecoli_dog_core_snps.fasta.treefile")
Tree$tip.label
Tree2 <- Tree
Tree2$tip.label <- substr(Tree$tip.label, 1, regexpr("\\.", Tree$tip.label)-1)
Tree2$tip.label
all(Tree2$tip.label %in% rownames(SNPs))
all(rownames(SNPs) %in% Tree2$tip.label)
all(Tree2$tip.label %in% rownames(SNPs))
all(rownames(SNPs) %in% Tree2$tip.label)
all(Tree2$tip.label %in% names(phen.Amikacin2))
all(names(phen.Amikacin2) %in% Tree2$tip.label)
all(names(phen.Amikacin2) %in% rownames(SNPs))
all(rownames(SNPs) %in% names(phen.Amikacin2))
names(phen.Amikacin2)

# suffixes <- keepLastN(colnames(SNPs), n = 2)
# suffixes <- unique(suffixes)

# if(all(suffixes %in% c(".a", ".c", ".g", ".t"))){
#   ## SNPs:
#   snps <- get.binary.snps(SNPs)
#   
#   ## ... Optional step (for snps.reconstruction, if present)
#   ## (Example only, if needed for user data)
#   # if(is.matrix(snps.reconstruction)){
#   #   snps.reconstruction <- snps.reconstruction[, which(colnames(snps.reconstruction) %in% colnames(snps))]
#   # }
# }


rm(eColiSNPs,eColiMat,eColiMatShrt,anno,headr,headr.want,anno.want,phenAll,eColirnames,eColirnames.sub,eColiMat2,Phen,phen.sub,phen.sub.0,phen.sub.01,phen.sub.01.1,phen.sub.01.2,phen.sub.01.3,phen.sub.01.4,phen.Amikacin,Tree)
ls()

treeWASAmikacin <- treeWAS(snps = SNPs,
                  phen = phen.Amikacin2,
                  tree = Tree2,
                  mem.lim = TRUE,
                  seed = 1)
```

```{r visualize_parse_data,include=TRUE}
# Will need to periodically update the .Rdata file from the server to get the most up-to-date results
sigSnpsTermLocus <- treeWASTest2$terminal$sig.snps$SNP.locus
write.table(sigSnpsTermLocus,file="/Users/hcm59/Box/Holly/Analyses/treeWAS/sigSNPsLociTerm_amikacin.txt")

print(treeWASTest2, sort.by.p=FALSE)
sig.loci <- treeWASTest2$treeWAS.combined
View(sig.loci)
sig.loci.indiv <- treeWASTest2$treeWAS.combined$treeWAS
sigSnpsTerm <- treeWASTest2$terminal$sig.snps
sigSnpsSimul <- treeWASTest2$simultaneous$sig.snps
sigSnpsSubse <- treeWASTest2$subsequent$sig.snps
write.csv(sigSnpsTerm,file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/sigSNPsTerm_amikacin.csv")

## No Significant SNPs found with simultaneous score 

plot_sig_snps(treeWASTest2$terminal$corr.dat,treeWASTest2$terminal$corr.sim, sig.snps = treeWASTest2$terminal$sig.snps)

phenRec <- treeWASTest2$dat$phen.reconstruction
snpRec <- treeWASTest2$dat$snps.reconstruction
phenTree <- treeWAS::plot_phen(Tree2, phen.nodes = phenRec, snp.nodes = snpRec[,1], par(pin=c(5,10)))

write.tree(phenTree,file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/testTree.nw")

# write.treeWAS(,filename="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/sigSNPsTerm_amikacin.csv")
```
## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
