---
title: "treeWAStake2"
output: html_document
---

```{r script, include=TRUE}
### Script for running treeWAS on the AHDC server

## First need to load in/install treeWAS
## install devtools, if necessary:
# install.packages("devtools", dep=TRUE)
library(devtools)

## install treeWAS from github:
# install_github("caitiecollins/treeWAS", build_vignettes = TRUE)
library(treeWAS)

## Read data from file:
eColiSNPs <- read.dna(file = "/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/results_with_long_branch_sequence_removed/Ecoli_dog_core_snps.fasta", format = "fasta")
# eColiSNPs <- read.dna(file = "/local/workdir/hcm59/Ecoli/R_analyses/Ecoli_dog_core_snps.fasta", format = "fasta")
rownames(eColiSNPs)
colnames(eColiSNPs)
str(eColiSNPs)
## Convert: 
eColiMat <- DNAbin2genind(eColiSNPs)@tab
dim(eColiMat)
# Remove columns with more than 50% NAs: 
eColiMatShrt <- eColiMat
eColiMatShrt2 <- eColiMatShrt[, which(colMeans(!is.na(eColiMatShrt)) > 0.5)]
dim(eColiMatShrt2)

# anno <- read.csv("/local/workdir/hcm59/Ecoli/R_analyses/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)
anno <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)
header <- colnames(anno)
header.want <- c(grep("Strain",header),grep("Assembly",header))
anno.want <- anno[header.want]
# View(anno.want)

# phenAll <- read.table("/local/workdir/hcm59/Ecoli/R_analyses/phenAll.txt")
phenAll <- read.table("/Users/hcm59/Box/Github/CornellPostdoc/phenAll.txt",header=TRUE)
eColirnames <- rownames(eColiMatShrt2)
eColirnames
eColirnames.sub <- substr(eColirnames, 1, regexpr("\\.", eColirnames)-1)
str(eColirnames.sub)
typeof(eColirnames.sub)
eColiMat2 <- eColiMatShrt2
rownames(eColiMat2) <- eColirnames.sub

all(rownames(names(phenAll)) %in% rownames(eColiMat2))

Phen <- phenAll
headr <- colnames(Phen)
headr.want <- c(grep("sample.ID",headr),grep("Assembly",headr), grep("INT",headr))
headr.want
phen.sub <- Phen[headr.want]
head(phen.sub)
names(phen.sub)[2]
typeof(phen.sub[2])
typeof(phen.sub[1])

library(dplyr)
library(reshape2)
# phen.Melt <- melt(Phen, id.vars=c("Assembly"),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)


# find all instances of "SUSC" and replace with 0
library(tidyverse)
phen.sub.0 <- phen.sub %>% mutate_if(is.atomic,funs(str_replace(., "SUSC", "0")))
phen.sub.0
# worked! # now do that to RESIST and convert NOINTP
phen.sub.01 <- phen.sub.0 %>%  mutate_if(is.atomic,funs(str_replace(., "RESIST", "1")))
phen.sub.01.1 <- phen.sub.01 %>%  mutate_if(is.atomic,funs(str_replace(., "NOINTP", "")))
phen.sub.01.2 <- phen.sub.01.1 %>%  mutate_if(is.atomic,funs(str_replace(., "INTER", "")))
View(phen.sub.01.2)
phen.sub.01.3 <- phen.sub.01.2 %>% mutate_all(na_if,"")


rownames(phen.sub.01.3) <- phenAll$Assembly
rownames(phen.sub.01.3)
View(phen.sub.01.3)
phen.sub.01.4 <- phen.sub.01.3[,2:104]
View(phen.sub.01.4)
str(phen.sub.01.4)

# phen.melt <- melt(phen.sub.01.4, id.vars=c("Assembly"),variable.name = "antibiotic", value.name = "INT",na.rm = TRUE)
phen.Amikacin <- phen.sub.01.4[,2]
names(phen.Amikacin) <- rownames(phen.sub.01.4)
View(phen.Amikacin)
typeof(phen.Amikacin)
phen.Amikacin

phen.Amikacin2 <- as.vector(unlist(phen.Amikacin))
names(phen.Amikacin2) <- names(phen.Amikacin)
all(rownames(names(phen.Amikacin2)) %in% rownames(eColiMat2))
all(rownames(eColiMat2) %in% names(phen.Amikacin2))
## Need files: phenotypic file, SNP file, tree file
# phen_Ami <- read.table("/local/workdir/hcm59/Ecoli/R_analyses/phen_Amikacin.txt")
# phen.Ami <- as.vector(unlist(phen_Ami))
# names(phen.Ami) <- rownames(phen_Ami)
SNPs <- eColiMat2
# Tree <- read.tree("/local/workdir/hcm59/Ecoli/R_analyses/Ecoli_dog_core_snps.fasta.treefile")
Tree <- read.tree("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/results_with_long_branch_sequence_removed/Ecoli_dog_core_snps.fasta.treefile")
Tree$tip.label
Tree2 <- Tree
Tree2$tip.label <- substr(Tree$tip.label, 1, regexpr("\\.", Tree$tip.label)-1)
Tree2$tip.label
all(Tree2$tip.label %in% rownames(SNPs))
all(rownames(SNPs) %in% Tree2$tip.label)
all(Tree2$tip.label %in% rownames(SNPs))
all(rownames(SNPs) %in% Tree2$tip.label)
all(Tree2$tip.label %in% names(phen.Amikacin2))
all(names(phen.Amikacin2) %in% Tree2$tip.label)
all(names(phen.Amikacin2) %in% rownames(SNPs))
all(rownames(SNPs) %in% names(phen.Amikacin2))
names(phen.Amikacin2)

rm(eColiSNPs,eColiMat,eColiMatShrt,anno,headr,headr.want,anno.want,phenAll,eColirnames,eColirnames.sub,eColiMat2,Phen,phen.sub,phen.sub.0,phen.sub.01,phen.sub.01.1,phen.sub.01.2,phen.sub.01.3,phen.sub.01.4,phen.Amikacin,Tree,header,header.want)
ls()

suffixes <- keepLastN(colnames(SNPs), n = 2)
suffixes <- unique(suffixes)

if(all(suffixes %in% c(".a", ".c", ".g", ".t"))){
  ## SNPs:
  snps <- get.binary.snps(SNPs)
  
  ## ... Optional step (for snps.reconstruction, if present)
  ## (Example only, if needed for user data)
  # if(is.matrix(snps.reconstruction)){
  #   snps.reconstruction <- snps.reconstruction[, which(colnames(snps.reconstruction) %in% colnames(snps))]
  # }
}





treeWASTest2 <- treeWAS(snps = snps,
                  phen = phen.Amikacin2,
                  tree = Tree2,
                  mem.lim = 16,
                  filename.plot = "/Users/hcm59/Box/Github/CornellPostdoc/treeWas_plots3.18.pdf",
                  seed=1)

# plot_phen(Tree2, phen.nodes=phen.plot.col$all.nodes)

```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
