---
title: "galaxy output visualizations"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(ggpubr)
library(stringr)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r load_data, include=TRUE}
staramr_output <- read.delim(file="/Users/hcm59/Box/Holly/Dog_E_coli_project/Nov_Assembly_Download/staramr_11.15.21/detailed_summary.tsv",header=TRUE,sep="\t")
# ST 131 star amr output 
# staramr_output_ST131 <- read.delim(file="/Users/hcm59/Box/Holly/Dog_E_coli_project/Galaxy7666-[staramr_on_data_7636,_data_7637,_and_others__detailed_summary.tsv].txt",header=TRUE,sep="\t")

# want to get rid of everything after the first .1 so that I can combine datasets with this same type of identifier 
# data_clean_phrase <- c("GCA_012245225.1_PDT000477478.1_genomic.fna", 
                       # "GCA_012246585.1_PDT000477476.1_genomic.fna")
# out <- stringr::str_extract(data_clean_phrase, "^(.*?.1_)") # including the @
# this leaves me with a trailing underscore (_) - should be easy enough to get rid of 
# out.2 <- str_remove_all(out, "[_]")

# make this into a function so can use it with apply across the entire dataframe
A <- function(x) {
  x.1 <- stringr::str_extract(x, "^(.*?.1_)") 
  out.2 <- str_remove_all(x.1, "[_]")
  return(out.2)
}


staramr_output_clean <- data.frame( apply(staramr_output[1],2, A), staramr_output[2:11])
# staramr_output_clean_ST131 <- data.frame( apply(staramr_output_ST131[1],2, A), staramr_output_ST131[2:11])

# probably should separate out the different categories (MLST, plasmid, Resistance)
levels(as.factor(staramr_output_clean$Data.Type))
MLST <- subset(staramr_output_clean,Data.Type=="MLST")
plasmid <- subset(staramr_output_clean,Data.Type=="Plasmid")
write.csv(plasmid,file="/Users/hcm59/Box/Holly/Dog_E_coli_project/Nov_Assembly_Download/plasmid_staramr_all.csv")
resistance <- subset(staramr_output_clean,Data.Type=="Resistance")
write.csv(resistance,file="/Users/hcm59/Box/Holly/Dog_E_coli_project/Nov_Assembly_Download/resistance_staramr_all.csv")
nlevels(as.factor(MLST$Data))
nlevels(as.factor(plasmid$Data))
nlevels(as.factor(resistance$Data))
nlevels(as.factor(resistance$Isolate.ID))
write.csv(MLST,file="/Users/hcm59/Box/Holly/Dog_E_coli_project/Nov_Assembly_Download/MLST_all.csv")
  

# remove rows with empty isolate ID column 
cols_to_check = c("Isolate.ID")

library(dplyr)
resistance %>% 
  filter(if_all(cols_to_check, ~ !is.na(.x)))

library(tidyr)
# remove NAs from isolate ID col 
resistance_noNA <- resistance[!is.na(resistance$Isolate.ID),]
resistance_spread <- tidyr::pivot_wider(resistance_noNA,id_cols=Isolate.ID,names_from=Data,values_from=Predicted.Phenotype,names_repair="minimal")
# str(resistance_spread)
# make dataframe 
resistance_spread_df <- as.data.frame(resistance_spread)
na_code <- "NULL"
rownames(resistance_spread_df) <- resistance_spread_df$Isolate.ID
to_binary <- function(m){
  m2 <- apply(m, 2, function(x){ ifelse(x %in% na_code, 0, 1) } )
  colnames(m2) <- colnames(m)
  rownames(m2) <- rownames(m)
  return(m2)
}
test2 <- to_binary(resistance_spread_df)
as.data.frame(test2)
test2[,1] <- rownames(test2)
str(test2)

resistance_spread_df_cleaned <- test2
write.csv(resistance_spread_df_cleaned,file="/Users/hcm59/Box/Holly/Dog_E_coli_project/Nov_Assembly_Download/resistance_spread_dfClean.csv")
      
# yay that worked 

# now want to replace "NULL" with 0 and anything else with 1 
library(naniar)
test <- resistance_spread %>%  mutate_if(resistance_spread,funs(str_replace(., "NULL", "0")))

resistance_spread2 <- resistance_spread %>% mutate_all(na_if,"") 



```

```{r subsetting gff3s, include=FALSE,echo=FALSE}
# need to subset the gff3 files to include only the plasmid contigs for Roary 
# try using grep 
test <- read.table('/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMR_results/dog_verified_host/gff\ files/ST131/GCA_012257685.1.gff', sep="\t", quote="")
gff2 <- gff[grep("Supercontig_1.1", gff[,1]),]





```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
