---
title: "phenoGenoCorr"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(tidyr)



```


```{r phen,include=TRUE}
# First read in the phenotypic data from Olga (already subsetted from previous analysis with treeWAS)
# Phen <- read.table("/Users/hcm59/Box/Github/CornellPostdoc/phenAll.txt",header=TRUE)
# Next read in the genotypic data from AMR finder 
PRJNA324565 <- read.table("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMRFinder_results/ARG_table_20210305_123652_PRJNA324565.tsv",header=TRUE)
PRJNA318591 <- read.table("/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMRFinder_results/ARG_table_20210312_143707_PRJNA318591.tsv", header=TRUE)
## annotations file that contains IDs that match phenotype dataset and those that match SNP dataset 
anno <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)
phen2017 <- read.csv(file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/2017-ECOL-MIC-Vet-LIRN.csv",header=TRUE,na.strings = c("", "NA"))
phen2018 <- read.csv(file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/2018-ECOL-MIC-Vet-LIRN.csv",header=TRUE,na.strings = c("", "NA"))
phen2019 <- read.csv(file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/2019-ECOL-MIC-Vet-LIRN.csv",header=TRUE,na.strings = c("", "NA"))

```

```{r visualize_AMRFinder_data, include=TRUE}

accListPRJNA318591 <- read.table("/Users/hcm59/Downloads/PRJNA318591_SraAccList.txt")
accListPRJNA324565 <- read.table("")
nsamples91 <- nlevels(as.factor(PRJNA318591$Isolate))
# 198 samples found AMR genes
# 199 samples total 
nsamples65 <- nlevels(as.factor(PRJNA324565$Isolate))
# 101 samples found AMR genes 
# 157 samples total 
AMRFinder <- rbind(PRJNA324565,PRJNA318591)
199+157
198+101
356-299

# 356 samples total 

library(ggplot2)
library(tidyr)
library(tibble)
library(hrbrthemes)
library(dplyr)
## Let's try pHeatmap now 
library(pheatmap)

#remove duplicates 
# AMRFinder [!duplicated(AMRFinder[c(1,2)]),]
amrFinderByClass <- cbind(AMRFinder$Isolate,AMRFinder$Class)
amrFinderByClass <-as.data.frame(amrFinderByClass)
summary(amrFinderByClass)
test <- amrFinderByClass %>% group_by(V1,V2) %>% mutate(count = n())
test2 <- dcast(test, V1 ~ V2)
write.csv(test2, "/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/amr_genes_table_perSample.csv")

classes_per_sample <- c(7,1,1,1,1,1,1,1,3,1,1,1,1,1,1,1,6,1,1,1,1,10,1,1,7,1,1,1,1,1,1,1,1,1,1,1,1,1,2,10,1,1,1,1,1,6,4,1,1,1,3,6,1,1,1,1,1,1,1,1,5,1,3,6,1,1,4,8,1,1,1,1,1,1,3,1,1,3,1,6,1,1,9,1,1,1,1,2,1,5,2,1,1,2,1,7,1,7,1,7,1,2,1,1,1,2,4,1,7,1,8,5,11,5,5,10,4,4,1,1,1,1,1,1,5,2,1,4,7,4,7,1,1,7,1,1,1,1,3,1,1,1,1,1,1,1,1,1,1,1,1,1,1,5,2,1,3,1,1,1,1,1,1,1,5,1,1,1,1,1,1,9,1,1,8,1,1,1,1,1,1,1,4,1,1,3,1,1,6,1,1,1,1,1,6,7,8,4,6,8,2,1,6,1,4,1,1,1,5,1,9,9,7,1,1,1,1,1,1,3,1,4,1,7,1,6,1,1,1,1,1,9,1,1,1,1,3,6,1,1,1,1,2,1,1,1,1,1,4,1,2,1,1,1,1,1,1,1,1,1,5,1,1,1,1,1,1,1,1,1,1,1,1,1,2,1,1,1,5,1,1,1,3,1,4,1,1,1,1,1,1,1,1,6,1,1,1,1,3)
numSamplesNumClasses <- table(classes_per_sample)
str(numSamplesNumClasses)
numSamplesNumClasses <- as.data.frame(numSamplesNumClasses)

AMRclassesPerSample <- ggplot(data = numSamplesNumClasses, mapping = aes(x=classes_per_sample,y=Freq)) + geom_col(fill="purple") +
  labs(x='AMR Classes per Sample',y='# samples') +
  theme_classic()
pdf(file="/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMRFinder_results/AMRclassesperSample.pdf")
AMRclassesPerSample

## Want to know how often each sample is represented in dataset
samples <- AMRFinder$Isolate
genes_per_sample <- table(as.factor(samples))
genes_per_sample <- as.data.frame(genes_per_sample)
hist(genes_per_sample$Freq)
plot(genes_per_sample)

# set up cut-off values 
breaks <- c(1,2,5,10,15,20,25)
# specify interval/bin labels
tags <- c("[1)", "[2-5)", "[5-10)", "[10-15)", "[15-20)", "[20-25)")
# bucketing values into bins
group_tags <- cut(genes_per_sample$Freq, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)
# inspect bins
summary(group_tags)










# set up cut-off values 
breaks <- c(1,2,5,10,15,20,25)
# specify interval/bin labels
tags <- c("[1)", "[2-5)", "[5-10)", "[10-15)", "[15-20)", "[20-25)")
# bucketing values into bins
group_tags <- cut(genes_per_sample$Freq, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)
# inspect bins
summary(group_tags)

AMRgenesperSample <- ggplot(data = as_tibble(group_tags), mapping = aes(x=value)) + 
  geom_bar(stat="count",fill="turquoise",color="white",alpha=0.7) + 
  labs(x='AMR Genes per Sample',y='# samples') +
  theme_classic() + stat_count(geom = "text", colour = "black", size = 3.5,
aes(label = ..count..,vjust=-0.5))
pdf(file="/Users/hcm59/Box/Goodman\ Lab/Projects/bacterial\ genomics/Ecoli_dog_AMRFinder_results/AMRgenesperSample.pdf")
AMRgenesperSample




```




```{r reorganizing Phenotypic data,include=FALSE}
# PRJNA324565$Isolate == anno$Run 
# Phen$Strain == anno$Strain 
# ^ going to take a couple of steps to get phenotypic and genotypic data with the same names
## annotations file that contains IDs that match phenotype dataset and those that match SNP dataset 
anno <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/ecoli_dog_200contigs_w_panaroo_tag.csv",header=TRUE)

# want strain (ID used in phenotypic data) and Run (ID used in Geno data)
header <- colnames(anno)
header.want <- c(grep("Strain",header),grep("Run",header))
anno.want <- anno[header.want]
View(anno.want)

# merge phenotypic data with annotations to get IDs and assembly in same file 
library(dplyr)
colnames(phen2017)
phen2017.anno <- inner_join(anno.want, phen2017, by=c("Strain" = "sample.ID."),keep=TRUE)
View(phen2017.anno)
headr2017 <- colnames(phen2017.anno)
headr2017.want <- c(grep("Strain",headr2017),grep("Run",headr2017), grep("sample.ID.",headr2017),grep("INT",headr2017))
                    # ,grep(".MIC",headr2017)
headr2017.want 
phen2017.anno2 <- phen2017.anno[headr2017.want]
dim(phen2017.anno2)
# rm(headr2017,headr2017.want)

phen2018.anno <- inner_join(anno.want, phen2018, by=c("Strain" = "sample.ID."),keep=TRUE)
View(phen2018.anno)
headr2018 <- colnames(phen2018.anno)
headr2018.want <- c(grep("Assembly",headr2018),grep("Run",headr2018),grep("sample.ID.",headr2018),grep("INT",headr2018))
                    # ,grep("MIC",headr2018))
headr2018.want 
phen2018.anno <- phen2018.anno[headr2018.want]
dim(phen2018.anno)
# rm(headr2018,headr2018.want)

phen2019.anno <- inner_join(anno.want, phen2019, by=c("Strain" = "sample.ID."),keep=TRUE)
# View(phen2019.anno)
headr2019 <- colnames(phen2019.anno)
headr2019.want <- c(grep("Assembly",headr2019),grep("Run",headr2019),grep("sample.ID.",headr2019),grep("INT",headr2019))
                    # ,grep("MIC",headr2019))
headr2019.want 
phen2019.anno <- phen2019.anno[headr2019.want]
dim(phen2019.anno)
# rm(headr2019,headr2019.want)

# rm(phenAll)
phenAll <- dplyr::bind_rows(phen2017.anno,phen2018.anno,phen2019.anno)
headr <- colnames(phenAll)
headr.want <- c(grep("Strain",headr),grep("Run",headr),grep("INT",headr))
headr.want
phen.sub <- phenAll[headr.want]

# need to tidy up phenotypic data 
library(reshape2)
typeof(phen.sub)
testTidayPhen <- gather(phen.sub,Subclass,Phenotype,-Strain,-Run) 
tidyPhen <- testTidayPhen[,2:4]
#remove substrings (.INT)
tidyPhen$Subclass <- gsub(".INT", "", tidyPhen$Subclass)

```

```{r reorganizing  data, include=FALSE}
# get a combined set of genotypic data 
genotype9165 <- rbind(PRJNA324565,PRJNA318591)

levels(as.factor(genotype9165$Subclass))
levels(as.factor(genotype9165$Class))

##subset genotypic data into subclasses
# actually want to subset it into classes - and put the phenotype antibiotics into their respective class for correlations  

gtAminogly <- genotype9165[which(genotype9165$Class=="AMINOGLYCOSIDE"),]
gtAminoglyQuin <- genotype9165[which(genotype9165$Class=="AMINOGLYCOSIDE/QUINOLONE"),]
gtBetaLac <- genotype9165[which(genotype9165$Class=="BETA-LACTAM"),]
gtBleo <- genotype9165[which(genotype9165$Class=="BLEOMYCIN"),]
gtColis <- genotype9165[which(genotype9165$Class=="COLISTIN"),]
gtFos <- genotype9165[which(genotype9165$Class=="FOSFOMYCIN"),]
gtLin <- genotype9165[which(genotype9165$Class=="LINCOSAMIDE"),]
gtMac <- genotype9165[which(genotype9165$Class=="MACROLIDE"),]
gtPhen <- genotype9165[which(genotype9165$Class=="PHENICOL"),]
gtQuat <- genotype9165[which(genotype9165$Class=="QUATERNARY_AMMONIUM"),]
gtQuin <- genotype9165[which(genotype9165$Class=="QUINOLONE"),]
gtRif <- genotype9165[which(genotype9165$Class=="RIFAMYCIN"),]
gtStrepthri <- genotype9165[which(genotype9165$Class=="STREPTOTHRICIN"),]
gtSulfo <- genotype9165[which(genotype9165$Class=="SULFONAMIDE"),]
gtTetra <- genotype9165[which(genotype9165$Class=="TETRACYCLINE"),]
gtTrim <- genotype9165[which(genotype9165$Class=="TRIMETHOPRIM"),]


# subset resitant phenotypes and susceptible phenotypes
subclass <- levels(as.factor(tidyPhen$Subclass))
subclass
lookupTable <- read.csv("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/lookupTableEcoli.csv",header=TRUE)

library(devtools, quietly=TRUE)
source_gist("https://gist.github.com/dfalster/5589956")
# addNewData("dataNew.csv", myData, allowedVars)
allowedVars <- c("Run","Subclass","Phenotype","Class")
ptSub <- addNewData("/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/lookupTableEcoli.csv", tidyPhen, allowedVars)
levels(as.factor(ptSub$Class))
View(ptSub)

table(levels(as.factor(genotype9165$Class)) %in% levels(as.factor(ptSub$Class)))
table(levels(as.factor(ptSub$Class)) %in% levels(as.factor(genotype9165$Class)))
levels(as.factor(genotype9165$Class))
ptAMINOGLY <- ptSub[which(ptSub$Class=="AMINOGLYCOSIDE"),]
dim(ptAMINOGLY)
dim(gtAminogly)
allAMINOGLY <- full_join(ptAMINOGLY, gtAminogly, by=c("Run" = "Isolate"),keep=TRUE)
write.csv(allAMINOGLY,file="/Users/hcm59/Box/Holly/Analyses/Analyses/Ecoli/allAMINOGLY.csv")
levels(as.factor(allAMINOGLY$Subclass.x))
#combine datasets by Run (resistNa) and Isolate (genotype9165)
# Only certain antibiotics are shared by both datasets







```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
